### PR TITLE
nonsensical MUST on frames fitting in packets

### PR DESCRIPTION
**Pull request by martinthomson**
These were phrased as requirements, but they are really just statements
of fact.

### CODE REVIEW

**Comment by mikkelfj** [comment]
I'm not sure that I agree. You could easily have a long stream frame spilling over into the next packet. There is nothing but implied complications that really prevents that from happing, except from this text. That said, but old and new version will work.

**Comment by janaiyengar** [comment]
I've answered this very question, including recently, to people who were trying to understand QUIC framing. The text is definitely not nonsensical, and unless you want people experimenting with making frames go across packet boundaries (yes, you _can_ do this and use packet numbers as a way to reassemble pieces of a frame), I'd keep the MUST. I don't see a reason to drop the MUST.

This isn't editorial. The MUST is not implied anywhere.

**Comment by LPardue** [comment]
I'm in the other camp. The spec doesn't describe means for frames to span packets, so how does a receiving peer detect that this has occurred, and how would it react if it did?

**Comment by ianswett** [comment]
I'm still not understanding how you get a frame to span packet boundaries.  I mean, you could do that, but the first one would cause the connection to close and if a subsequent one arrived first who knows what it would be interpreted as.

**Comment by janaiyengar** [comment]
@ianswett : The first one would cause the connection to close because the receiver is enforcing a MUST.  It's not clear to me why requiring the same thing of the sender is a problem.

**Comment by janaiyengar** [comment]
@LPardue: We require the receiver to close the connection when it doesn't understand how to handle a frame that is split across packets, why not require the sender to not do this thing that will cause the receiver to close the connection.

I could be in the rough here, but I don't understand why we need to drop this requirement. I can live with either text, but I'd prefer the requirement.

**Comment by mikkelfj** [comment]
> I'm in the other camp. The spec doesn't describe means for frames to span packets, so how does a receiving peer detect that this has occurred, and how would it react if it did?

The fact that this is difficult does not prevent the sender from doing it. The receiver could detect that a stream frame was shorter that advertised and assume the next higher packet number must have the rest, or some of the rest. Of course, this is an incorrect assumption, but if the receivers implementation is the same as the senders, this could work for a while, until it breaks.

**Comment by kazuho** [comment]
I am favor of adopting the PR. Regarding @janaiyengar's concern, maybe we can rephrase the paragragh right before from:

_The payload of QUIC packets, after removing packet protection, commonly consists of a sequence of frames, as shown in Figure 7. Version Negotiation, Stateless Reset, and Retry packets do not contain frames._

to :

_The Payload of QUIC packets, after removing packet protection, consists of a sequence of complete frames, as shown in Figure 7. Version Negotiation, Stateless Reset, and Retry packets do carry the Payload._

Note that the three exceptional packets (VN, Stateless Reset, Retry) do not carry "Payload" by their definition.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2678,12 +2678,9 @@ Packet number encoding at a sender and decoding at a receiver are described in
 
 ## Frames and Frame Types {#frames}
 
-The payload of QUIC packets, after removing packet protection, commonly consists
-of a sequence of frames, as shown in {{packet-frames}}.  Version Negotiation,
-Stateless Reset, and Retry packets do not contain frames.
-
-<!-- TODO: Editorial work needed in this section. Not all packets contain
-frames. -->
+The payload of QUIC packets, after removing packet protection, commonly consist
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]:
```suggestion
The payload of a QUIC packet, after removing packet protection, commonly consists
```

or "payload of QUIC packets, ... consists"

**Comment by martinthomson** [comment]
@janaiyengar 
> The MUST is not implied anywhere.

It's implied everywhere.  It's implied by the fact that UDP is lossy.  It is implied in that there are rules for handling of malformed frames and mandates to generate a connection error in that case.

You would prefer that it be explicit, I get that, but this is better described as an axiom.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: ... maybe remove "commonly" as well, since the exceptions are listed next anyway?

**Comment by janaiyengar** [comment]
> You would prefer that it be explicit, I get that, but this is better described as an axiom.

I think that's where the disconnect was. @kazuho's definitional change made it clear to me that you were all seeing it as axiomatic, where I (and perhaps @mikkelfj as well) was seeing it as a choice that needed enforcement.  And making the definitional change makes the axiom explicit.

**Pull request merged**